src/adapter/src/frontend_peek.rs

```diff
@@ -374,137 +439,309 @@ impl PeekClient {
+                                }
+                            }
+                        }
+                        ExplainContext::None => {
+                            Ok(ExecuteOrExplain::Execute {
+                                global_lir_plan: global_lir_plan_result?,
+                                optimization_finished_at,
+                                plan_insights_optimizer_trace: None,
+                                insights_ctx: None,
+                            })
+                        }
+                        ExplainContext::PlanInsightsNotice(optimizer_trace) => {
+                            let global_lir_plan = global_lir_plan_result?;
+                            let insights_ctx = create_insights_ctx(&optimizer, true);
+                            Ok(ExecuteOrExplain::Execute {
+                                global_lir_plan,
+                                optimization_finished_at,
+                                plan_insights_optimizer_trace: Some(optimizer_trace),
+                                insights_ctx,
+                            })
+                        }
+                        ExplainContext::Pushdown => {
+                            // TODO(peek-seq): Decide whether we want EXPLAIN FILTER PUSHDOWN to be supported in the frontend sequencing. I don't see a reason a the moment.
+                            Err(AdapterError::Internal(
+                                "EXPLAIN FILTER PUSHDOWN not supported in frontend peek. Should have bailed out above already, because it's not EXPLAIN SELECT.".to_string()
+                            ))
+                        }
+                    }
                 })
             },
         )
         .await
-        {
-            Ok(r) => r,
-            Err(adapter_error) => {
-                return Err(adapter_error);
+        .map_err(|join_error| AdapterError::Internal(format!("optimization task panicked: {}", join_error)))?;
+
+        // Handle the optimization result: either generate EXPLAIN output or continue with execution
+        match optimization_result {
+            ExecuteOrExplain::Explain {
+                df_meta,
+                explain_ctx,
+                optimizer,
+                insights_ctx,
+            } => {
+                // Generate EXPLAIN response
+                Ok(Some(
+                    self.generate_explain_response(
+                        session,
+                        &catalog,
+                        df_meta,
+                        explain_ctx,
+                        optimizer,
+                        insights_ctx,
+                    )
+                    .await?,
+                ))
             }
-        };
-
-        // # From peek_finish
+            ExecuteOrExplain::Execute {
+                global_lir_plan,
+                optimization_finished_at: _optimization_finished_at,
+                plan_insights_optimizer_trace,
+                insights_ctx,
+            } => {
```

Comment:
This whole block got an extra indentation, so it's better to view this with whitespace changes hidden.